client/extensions/woocommerce/state/sites/product-categories/reducer.js

```diff
@@ -1,50 +1,120 @@
-			return LOADING;
-		},
+/**
+ * Returns if a product categories request for a specific query is in progress or not.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function isQueryLoading( state = {}, action ) {
+	switch ( action.type ) {
+		case WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST:
+		case WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST_SUCCESS:
+		case WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST_FAILURE:
+			const query = getSerializedProductCategoriesQuery( action.query );
+			return Object.assign( {}, state, {
+				[ query ]: WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST === action.type,
+			} );
+		default:
+			return state;
+	}
+}
+
+/**
+ * Returns if a product categories request for a specific query has returned an error.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function isQueryError( state = {}, action ) {
+	if ( WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST_FAILURE === action.type && action.error ) {
+		const query = getSerializedProductCategoriesQuery( action.query );
+		return Object.assign( {}, state, { [ query ]: true } );
+	}
 
-		[ WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST_SUCCESS ]: ( state, { data } ) => {
-			return data;
-		},
+	return state;
+}
 
-		[ WOOCOMMERCE_PRODUCT_CATEGORY_UPDATED ]: productCategoryUpdated,
+/**
+ * Tracks all known categories objects, indexed by ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function items( state = {}, action ) {
+	if ( WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST_SUCCESS === action.type && action.data ) {
+		const cats = keyBy( action.data, 'id' );
+		return Object.assign( {}, state, cats );
+	}
+
+	if ( WOOCOMMERCE_PRODUCT_CATEGORY_UPDATED === action.type && action.data ) {
+		return {
+			...state,
+			[ action.data.id ]: action.data,
+		};
 	}
-);
 
-function productCategoryUpdated( state, action ) {
-	const { data } = action;
-	const productCategories = state !== LOADING ? state : [];
-	return updateCachedProductCategory( productCategories, data );
+	return state;
 }
 
-function updateCachedProductCategory( productCategories, category ) {
-	let found = false;
-	const newCategories = productCategories.map( c => {
-		if ( c.id === category.id ) {
-			found = true;
-			return category;
-		}
-		return c;
-	} );
-
-	if ( ! found ) {
-		newCategories.push( category );
+/**
+ * Tracks the categories which belong to a query, as a list of IDs
+ * referencing items in `productCategories.items`.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function queries( state = {}, action ) {
+	if ( WOOCOMMERCE_PRODUCT_CATEGORIES_REQUEST_SUCCESS === action.type && action.data ) {
+		const idList = action.data.map( cat => cat.id );
```

Comment:
Is there any chance we could potentially have duplicate category ids somehow? maybe during addition of many categories and certain ID's moving between pages of results... would it hurt to `uniq` the array before we persist `idList`? Or am I just being :paranoid:?

Reply:
I think there is the potential for that if someone adds a bunch of categories between page requests like you said, but I'm not sure if we really need to worry about it. When switching between pages in product reviews for example, the list is refreshed to get the latest IDs so the correct results get updated as they page around. I feel like adding `uniq` would complicate this a bit, since you would potentially need to start refreshing the other queries.